### ISSUE TITLE
GSoC 2017 proposal : Complete the implementation of FrozenSet, Set and Str builtin data types. Partial implementation of JSON module

### ISSUE THREAD

**Issue by ASP1234** [pybee/voc#487]
### Abstract/summary: 

VOC is a transpiler that converts Python code into Java class files so Python code can run in the JVM. This enables you to run Python code anywhere that there is a JVM - which includes Android phones, web applications to deploy in a J2EE container. 

In order to replicate Python behavior in Java, Python's logic for all the basic operations needs to be implemented in the VOC Java support library.
### Describe the need your project fulfills: 

According to Effective Java, one of the idioms of a good software is to be Immutable whenever possible :

 - It’s easier to reason about immutable code
 - It’s easier to prevent accidental changes if the data is immutable
 - It’s easier to execute code in parallel when the data is immutable
 
A frozenset is an immutable set - it's set up with the frozenset() function call, but once it's set up its contents cannot be altered. Operations which are non-intrusive (i.e. read only) work on a frozenset in the same way that they work on sets, but of course anything that writes to / modifies a set can't be used on a frozenset.

One of the most widely used data types is a string. A string consists of one or more characters, which can include letters, numbers, or other types of characters. So, a complete implementation of String datatype is crucial for this project.

JSON is JavaScript Object Notation. It is a much-more compact way of transmitting sets of data across network connections as compared to XML and thus, widely preferred and used.

### How will your project meet this need: 

 - **Set Types -- set, frozenset**

I have sub grouped each of the operations into two categories :

	 - Set Operations:
This includes *add, clear, copy, difference, difference_update, discard, remove, intersection, isdisjoint, issubset, issuperset, pop*. Java Collections can be used here. The main challenge is to deal with immutability in the implementation of frozenset operations. Also, the sweetness lies in the fact that Set and FrozenSet resemble in many ways. So, I'll be tackling both simultaneously.

	 - Built-in Functions:
The Python interpreter has a number of [functions](https://docs.python.org/3/library/functions.html) built into it that are always available.  Output of Python code doing that operation should be the same as the one run through CPython.

 - **String** 

I have sub grouped each of the operations into two categories :

	 - Complex Operations:
These operations don't have any naive Java implementation or require additional logic.
This includes *translate, maketrans, format, format_map*. 

	 - Java Support Functions:
This includes *built-in functions, unicode operations, splitlines*. 

- **JSON** 

The two major operations of JSON:

	 - Encode:
 Encodes the Python object into a JSON string representation. 

	 - Decode:
Decodes a JSON-endoded string into a Python object.

For this, we have JSON.Simple Library available to parse the JSON Strings in Java. But then this means introducing an external dependency into the project. So, either we need to implement a parser of our own or can rely on JSON.Simple.

### Timeline/milestones:

**Note: The whole process will be accomplished in stages, so that rebasing and merging doesn't pose any difficulties.**

##### **May 4 2017 - May 30, 2017** [Community Bonding]

* Gain greater insight into JSON codebase
* Bond with the community members and become accustomed with the Beeware culture
* Learn the tools and technologies necessary for the project
* Start a blog about the project to help my mentor track my progress

##### **May 30, 2017 - June 02, 2017**

 - Mantra is to implement operations in Set and modify it for FrozenSet. This ensures lesser bugs and better maintainability
 - Implement the *difference, difference_update, intersection and intersection_update* methods in FrozenSet
 - Implement *isdisjoint, issubset and issuperset* in Set

##### **June 05, 2017 - June 09, 2017**
 - Implement *isdisjoint, issubset, issuperset, union and update* in FrozenSet
 - Implement *symmetric_difference and symmetric_difference_update* in Set

##### **June 12, 2017 - June 16, 2017** 
- Implement *symmetric_difference and symmetric_difference_update* in FrozenSet
- Implement *format* in Set and FrozenSet

##### **June 19, 2017 - June 23, 2017** 

 - Implement *Built-in Functions* for Set and FrozenSet
 - Any bugs encountered to this point or project goals not implemented yet to be tackled by this time.

##### **June 26, 2017 - June 30, 2017** [Phase 1 Evaluation]

* Mentors and students submit their evaluations of one another. These evaluations are a required step of the program.
* Implement  *str.isidentifier, str.isnumeric, str.isprintable*

##### **July 3, 2017 - July 7, 2017**

 - Implement *str.casefold, str.splitlines*
 - Implement *str.fromat, str.format_map*

##### **July 10, 2017 - July 14, 2017**
- Implement *str.translate, str.maketrans*
- Deep Dive into JSON module

##### **July 17, 2017 - July 21, 2017**
- Implement *str.encode*, *naive* methods
- Any bugs encountered to this point or project goals not implemented yet to be tackled by this time.
- Finalize library or scratch implementation for JSON module

##### **July 24, 2017 - July 28, 2017** [Phase 2 Evaluation]
 * Mentors and students submit their evaluations of one another. These evaluations are a required step of the program.
 * Implement dumps for JSON module
 * Start working on JSON Array Parsing
 
##### **July 31, 2017 - Aug 4, 2017**
 * Implement JSON decoder for JSON Array
 * Start working on JSON Object Parsing
 * Any bugs encountered to this point or project goals not implemented yet to be tackled by this time

##### **Aug 7, 2017 - Aug 11, 2017**

 - Implement JSON decoder for JSON Object
 - Implement loads for JSON module
 - Replicate the methodology for Encoder

##### **Aug 14, 2017 - Aug 18, 2017**
- Implement JSON encoder
-   Any bugs encountered to this point or project goals not implemented yet to be tackled by this time

##### **Aug 21, 2017 - Aug 29, 2017**
* Submit sample code to the mentors.
* Mentors review student code samples and determine if the students have successfully completed their Google Summer of Code 2017 project.

**************************

### What broader goal is your project working towards? 

Making VOC more genial and accessible to new and existing users and expand its use cases.

### What resources will you need: people, documentation, literature, sample data, hardware if applicable: 

* Community Development team's help to choose most suitable path for JSON module

* Community support and report issues if encountered related to my work

### Setup

* [Forked repository](https://github.com/ASP1234/voc.git)
* [Development environment](http://voc.readthedocs.io/en/latest/internals/contributing.html)

### Context

I'd be elated if the proposal could be accepted, as it could allow me to make more advanced contributions to the open source. The idea of contributing to the open source and helping the non - technical people through coding and problem - solving skills is exciting.

### Ongoing involvement

I intend to regularly contribute to the [VOC](https://github.com/pybee/voc.git) codebase and be involved in the gitter channel. I have these [contributions](https://github.com/pybee/voc/commits?author=ASP1234) to [VOC](https://github.com/pybee/voc.git) repository. I will continue contributing to [VOC](https://github.com/pybee/voc.git) codebase even after the GSOC 2017. I intend to stay around and be a active member for Pybee after the summer is over.

### Commitment

I understand that this is a serious commitment. I don't have any other major commitments and I will devote ~40 hours a week during the entire tenure.

**Comment by freakboy3742** [comment]
Thanks for submitting this proposal. I have two major comments:

1. I'm not convinced you've done enough exploration of the amount of work involved for each method. For example, in the week of June 26, you're proposing to implement str.isidentifier, str.isnumeric, and  str.isprintable - three relatively straightforward boolean checks. The following week, you're proposing to implement str.casefold, str.splitlines, str.fromat, and str.format_map - four *significantly* more complex methods. A similar problem is evident in the weeks of June 5 and June 12 - FrozenSet and Set will be mostly identical in implementation - yet you've allocated the same amount of time to implement the same two methods on both classes. The timeline you've proposed feels like you've determined what you want to do, and then split it up into 12 weeks, rather than working out how much time each task will take. 

2. You've suggested using JSON.Simple as a library. That's fine - there's no need to re-invent the JSON parsing wheel. But why is JSON.Simple the right library to use? What alternatives have you investigated?

**Comment by freakboy3742** [comment]
Unfortunately, this proposal wasn't accepted.
